src/Workspaces/Core/Portable/FindSymbols/FindReferences/FindReferencesSearchEngine.SymbolSet.cs

```diff
@@ -0,0 +1,271 @@
+            {
+                foreach (var symbol in symbols)
+                    stack.Push(symbol);
+            }
+
+            protected static async Task AddCascadedAndLinkedSymbolsToAsync(
+                FindReferencesSearchEngine engine, ImmutableArray<ISymbol> symbols, HashSet<ISymbol> to, Stack<ISymbol> workQueue, CancellationToken cancellationToken)
+            {
+                foreach (var symbol in symbols)
+                    await AddCascadedAndLinkedSymbolsToAsync(engine, symbol, to, workQueue, cancellationToken).ConfigureAwait(false);
+            }
+
+            protected static async Task AddCascadedAndLinkedSymbolsToAsync(
+                FindReferencesSearchEngine engine, ISymbol symbol, HashSet<ISymbol> to, Stack<ISymbol> workQueue, CancellationToken cancellationToken)
+            {
+                foreach (var finder in engine._finders)
+                {
+                    var cascaded = await finder.DetermineCascadedSymbolsAsync(symbol, engine._solution, engine._options, cancellationToken).ConfigureAwait(false);
+                    foreach (var cascade in cascaded)
+                        await AddLinkedSymbolsToAsync(cascade).ConfigureAwait(false);
+                }
+
+                await AddLinkedSymbolsToAsync(symbol).ConfigureAwait(false);
+
+                return;
+
+                // Adds all the symbols 'lniked' to <paramref name="symbol"/> in the given. The symbols found are added
+                // to <paramref name="to"/>.  If <paramref name="to"/> did not contain that symbol, then it is also
+                // added to <paramref name="workQueue"/> to allow fixed point algorithms to continue.
+                async Task AddLinkedSymbolsToAsync(ISymbol symbol)
+                {
+                    foreach (var linked in await SymbolFinder.FindLinkedSymbolsAsync(symbol, engine._solution, cancellationToken).ConfigureAwait(false))
+                    {
+                        if (to.Add(linked))
+                            workQueue.Push(linked);
+                    }
+                }
+            }
+
+            private static bool InvolvesInheritance(ISymbol symbol)
+                => symbol.Kind is SymbolKind.Method or SymbolKind.Property or SymbolKind.Event;
+
+            /// <summary>
+            /// Finds all the symbols 'down' the inheritance hierarchy of <paramref name="symbol"/> in the given
+            /// project.  The symbols found are added to <paramref name="to"/>.  If <paramref name="to"/> did not
+            /// contain that symbol, then it is also added to <paramref name="workQueue"/> to allow fixed point
+            /// algorithms to continue.
+            /// </summary>
+            protected async Task AddDownSymbolsAsync(
+                ISymbol symbol, HashSet<ISymbol> to, Stack<ISymbol> workQueue,
+                ImmutableHashSet<Project> projects, CancellationToken cancellationToken)
+            {
```

Comment:
this is logic htat used to be in the AbstractMethodOrPropertyOrEventReferenceFinder.  Instead of being there, all inheritance handling is just pulled up to the core engine itself.